drivers/docker/driver.go

```diff
@@ -0,0 +1,1682 @@
+func (d *Driver) loadImage(task *drivers.TaskConfig, driverConfig *TaskConfig, client *docker.Client) (id string, err error) {
+
+	archive := filepath.Join(task.TaskDir().LocalDir, driverConfig.LoadImage)
+	d.logger.Debug("loading image from disk", "archive", archive)
+
+	f, err := os.Open(archive)
+	if err != nil {
+		return "", fmt.Errorf("unable to open image archive: %v", err)
+	}
+
+	if err := client.LoadImage(docker.LoadImageOptions{InputStream: f}); err != nil {
+		return "", err
+	}
+	f.Close()
+
+	dockerImage, err := client.InspectImage(driverConfig.Image)
+	if err != nil {
+		return "", recoverableErrTimeouts(err)
+	}
+
+	coordinator, callerID := d.getDockerCoordinator(client, task)
+	coordinator.IncrementImageReference(dockerImage.ID, driverConfig.Image, callerID)
+	return dockerImage.ID, nil
+}
+
+func (d *Driver) containerBinds(task *drivers.TaskConfig, driverConfig *TaskConfig) ([]string, error) {
+
+	allocDirBind := fmt.Sprintf("%s:%s", task.TaskDir().SharedAllocDir, task.Env[env.AllocDir])
+	taskLocalBind := fmt.Sprintf("%s:%s", task.TaskDir().LocalDir, task.Env[env.TaskLocalDir])
+	secretDirBind := fmt.Sprintf("%s:%s", task.TaskDir().SecretsDir, task.Env[env.SecretsDir])
+	binds := []string{allocDirBind, taskLocalBind, secretDirBind}
+
+	if !d.config.VolumesEnabled && driverConfig.VolumeDriver != "" {
+		return nil, fmt.Errorf("'volumes_enabled' is false; cannot use volume driver %q", driverConfig.VolumeDriver)
+	}
+
+	for _, userbind := range driverConfig.Volumes {
+		parts := strings.Split(userbind, ":")
+		if len(parts) < 2 {
+			return nil, fmt.Errorf("invalid docker volume: %q", userbind)
+		}
+
+		// Resolve dotted path segments
+		parts[0] = filepath.Clean(parts[0])
+
+		// Absolute paths aren't always supported
+		if filepath.IsAbs(parts[0]) {
+			if !d.config.VolumesEnabled {
+				// Disallow mounting arbitrary absolute paths
+				return nil, fmt.Errorf("'volumes_enabled' is false; cannot mount host paths: %+q", userbind)
+			}
+			binds = append(binds, userbind)
+			continue
+		}
+
+		// Relative paths are always allowed as they mount within a container
+		// When a VolumeDriver is set, we assume we receive a binding in the format volume-name:container-dest
+		// Otherwise, we assume we receive a relative path binding in the format relative/to/task:/also/in/container
+		if driverConfig.VolumeDriver == "" {
+			// Expand path relative to alloc dir
+			parts[0] = filepath.Join(task.TaskDir().Dir, parts[0])
+		}
+
+		binds = append(binds, strings.Join(parts, ":"))
+	}
+
+	if selinuxLabel := d.config.VolumesSelinuxLabel; selinuxLabel != "" {
+		// Apply SELinux Label to each volume
+		for i := range binds {
+			binds[i] = fmt.Sprintf("%s:%s", binds[i], selinuxLabel)
+		}
+	}
+
+	return binds, nil
+}
+
+func (d *Driver) createContainerConfig(task *drivers.TaskConfig, driverConfig *TaskConfig,
+	imageID string) (docker.CreateContainerOptions, error) {
+
+	logger := d.logger.With("task_name", task.Name)
```

Comment:
Seems like we should be using loggers with the task name context a _lot_ more places... In hindsight we probably should have made all of the driver impls lightweight shims around task-specific structs/funcs with a contextual logger.

Definitely out of scope for this PR though. :)